### PR TITLE
Replace HMAC Python implementation with C

### PR DESCRIPTION
**Pull request by prusnak**
I wrapped HMAC code from trezor-crypto into modtrezorcrypto-hmac similarly to what we do with PBKDF2.

Syntax changes slightly from

```
hmac.new(key, msg, hashlib.sha256)
hmac.Hmac(key, msg, hashlib.sha512)
```

to

```
hmac(hmac.SHA256, key, msg)
hmac(hmac.SHA512, key, msg)
```

I kept pure-Python implementation of HMAC for Monero in `core/src/apps/monero/xmr/crypto/__init__.py` - please pay special attention to reviewing that.

### CODE REVIEW

File: core/src/apps/monero/xmr/crypto/__init__.py [hunk 1]
```diff
@@ -37,9 +37,19 @@ def keccak_2hash(inp, buff=None):
...
+    block_size = inner.block_size
+    if len(key) > block_size:
+        key = digestmod(key).digest()
+    key = key + bytes(block_size - len(key))
+    inner.update(bytes((x ^ 0x36) for x in key))
+    inner.update(msg)
+    outer = digestmod()
+    outer.update(bytes((x ^ 0x5C) for x in key))
+    outer.update(inner.digest())
+    return outer.digest()
```
Comment (matejcik) [inline core/src/apps/monero/xmr/crypto/__init__.py:52]: the `bytes(bla for bla in foo)` are huge in terms of allocation count
something like this is better (untested):

```suggestion
def compute_hmac(key, msg):
    digestmod = keccak_factory
    inner = digestmod()
    block_size = inner.block_size
    if len(key) > block_size:
        key = digestmod(key).digest()
    key_block = bytearray(block_size)
    for i in range(block_size):
        key_block[i] = 0x36
    for i in range(len(key)):
    	key_block[i] ^= key[i]
    inner.update(key_block)
    inner.update(msg)
    outer = digestmod()
    for i in range(block_size):
        key_block[i] = 0x5C
    for i in range(len(key)):
    	key_block[i] ^= key[i]
    outer.update(key_block)
    outer.update(inner.digest())
    return outer.digest()
```

**Comment by matejcik** [comment]
utACK, except for the change in Monero

**Comment by prusnak** [comment]
Accepted @matejcik's suggestion as a fixup commit 54e8f04e1e07d3da52c712ea3c166afc6455ab7c - let's see what monero test in the CI thinks about that :)

**Comment by prusnak** [comment]
Monero test worked with the fixup 54e8f04e1e07d3da52c712ea3c166afc6455ab7c, so I am rebasing this and autosquashing the fixup commit.

**Review by andrewkozlik — COMMENTED** [review]
In terms of optimizations it would be nice to have a function like https://docs.python.org/3/library/hmac.html#hmac.digest which calls the "inlined" `hmac_sha256()` from `hmac.c`, because most of the uses of hmac in our code are of the form `hmac(...).digest()`. However I am not convinced that this function is a major bottleneck, so it's probably not worth the trouble.

File: core/embed/extmod/modtrezorcrypto/modtrezorcrypto-hmac.h [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+///     """
+STATIC mp_obj_t mod_trezorcrypto_Hmac_digest(mp_obj_t self) {
+  mp_obj_Hmac_t *o = MP_OBJ_TO_PTR(self);
+  if (o->hashtype == SHA256) {
+    HMAC_SHA256_CTX ctx = {0};
+    memcpy(&ctx, &(o->ctx256), sizeof(HMAC_SHA256_CTX));
+    uint8_t out[SHA256_DIGEST_LENGTH] = {0};
+    hmac_sha256_Final(&ctx, out);
+    memzero(&ctx, sizeof(HMAC_SHA256_CTX));
+    return mp_obj_new_bytes(out, sizeof(out));
```
Comment (andrewkozlik) [inline core/embed/extmod/modtrezorcrypto/modtrezorcrypto-hmac.h:124]:
```suggestion
    vstr_t vstr = {0};
    vstr_init_len(&vstr, SHA256_DIGEST_LENGTH);
    hmac_sha256_Final(&ctx, (uint8_t *)vstr.buf);
    memzero(&ctx, sizeof(HMAC_SHA256_CTX));
    return mp_obj_new_str_from_vstr(&mp_type_bytes, &vstr);
```

Untested. Note that the same applies to the SHA-512 branch below.

BTW `hmac_sha256_Final()` already calls `memzero` on the context, so technically it's not needed here.

File: core/embed/extmod/modtrezorcrypto/modtrezorcrypto-hmac.h [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+    mp_raise_ValueError("Invalid hashtype");
+  }
+  return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_2(mod_trezorcrypto_Hmac_update_obj,
+                                 mod_trezorcrypto_Hmac_update);
+
+/// def digest(self) -> bytes:
+///     """
+///     Retrieve the digest key.
```
[the remark this answers is not in the document]
Comment (prusnak) [reply core/embed/extmod/modtrezorcrypto/modtrezorcrypto-hmac.h]: Addressed in 046c06113857474cf654734fc8277d45bf9bdd43

File: core/embed/extmod/modtrezorcrypto/modtrezorcrypto-hmac.h [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+  }
+  if (o->hashtype == SHA512) {
+    HMAC_SHA512_CTX ctx = {0};
+    memcpy(&ctx, &(o->ctx512), sizeof(HMAC_SHA512_CTX));
+    uint8_t out[SHA512_DIGEST_LENGTH] = {0};
+    hmac_sha512_Final(&ctx, out);
+    memzero(&ctx, sizeof(HMAC_SHA512_CTX));
+    return mp_obj_new_bytes(out, sizeof(out));
+  }
+  return mp_const_none;
```
[the remark this answers is not in the document]
Comment (prusnak) [reply core/embed/extmod/modtrezorcrypto/modtrezorcrypto-hmac.h:128]: Addressed in 046c06113857474cf654734fc8277d45bf9bdd43

File: core/embed/extmod/modtrezorcrypto/modtrezorcrypto-hmac.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (prusnak) [reply core/embed/extmod/modtrezorcrypto/modtrezorcrypto-hmac.h:117]: I agree, but see below

Comment (prusnak) [reply core/embed/extmod/modtrezorcrypto/modtrezorcrypto-hmac.h:117]: This seems fine, but we use `mp_obj_new_bytes` all over the modtrezorcrypto. I prefer consistency. If we think this should be changed (and I am not opposed), let's address this in another PR where we replace all calls to `mp_obj_new_bytes` with `mp_obj_new_str_from_vstr`.

Comment (andrewkozlik) [reply core/embed/extmod/modtrezorcrypto/modtrezorcrypto-hmac.h:117]: There is nothing wrong with using `mp_obj_new_bytes()` as long as we clean up the stack, which we don't do here.

The `aes`, `bip32`, `chacha20poly1305`, `random` and `shamir` modules use `vstr`, but you are right that the hash-related modules and some other important ones don't. I will create a PR for that.

File: core/embed/extmod/modtrezorcrypto/modtrezorcrypto-hmac.h [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+    HMAC_SHA512_CTX ctx = {0};
+    memcpy(&ctx, &(o->ctx512), sizeof(HMAC_SHA512_CTX));
+    uint8_t out[SHA512_DIGEST_LENGTH] = {0};
+    hmac_sha512_Final(&ctx, out);
+    memzero(&ctx, sizeof(HMAC_SHA512_CTX));
+    return mp_obj_new_bytes(out, sizeof(out));
+  } else {
+    mp_raise_ValueError("Invalid hashtype");
+  }
+  return mp_const_none;
```
Comment (andrewkozlik) [inline core/embed/extmod/modtrezorcrypto/modtrezorcrypto-hmac.h:128]: Here is an untested proposal for a change corresponding in style to the PR I just created:
```suggestion
  vstr_t mac = {0};
  if (o->hashtype == SHA256) {
    HMAC_SHA256_CTX ctx = {0};
    memcpy(&ctx, &(o->ctx256), sizeof(HMAC_SHA256_CTX));
    vstr_init_len(&mac, SHA256_DIGEST_LENGTH);
    hmac_sha256_Final(&ctx, (uint8_t *)mac.buf);
    memzero(&ctx, sizeof(HMAC_SHA256_CTX));
  } else if (o->hashtype == SHA512) {
    HMAC_SHA512_CTX ctx = {0};
    memcpy(&ctx, &(o->ctx512), sizeof(HMAC_SHA512_CTX));
    vstr_init_len(&mac, SHA512_DIGEST_LENGTH);
    hmac_sha512_Final(&ctx, (uint8_t *)mac.buf);
    memzero(&ctx, sizeof(HMAC_SHA512_CTX));
  } else {
    mp_raise_ValueError("Invalid hashtype");
  }
  return mp_obj_new_str_from_vstr(&mp_type_bytes, &mac);
```

**Review by matejcik — APPROVED** [review]

**Review by andrewkozlik — APPROVED** [review]

**Review by onvej-sl — APPROVED** [review]

**Pull request merged**